File: src/adapters/_shared.ts
import type {
  AgentArtifact,
  AgentArtifactInput,
  HookEntry,
  ManifestComponentPaths,
  PluginMetadata,
  PortableHooks,
} from "../domain/harness.js";
import { PORTABLE_EVENTS } from "../domain/harness.js";

const PASCAL_MAP: Record<(typeof PORTABLE_EVENTS)[number], string> = {
  sessionStart: "SessionStart",
  preToolUse: "PreToolUse",
  postToolUse: "PostToolUse",
};

const DEFAULT_HOOK_TIMEOUT = 600;

type ManifestComponentKey = keyof ManifestComponentPaths;

function pickManifestPaths(
  componentPaths: ManifestComponentPaths,
  supportedKeys: readonly ManifestComponentKey[],
): Record<string, string> {
  return Object.fromEntries(
    supportedKeys.flatMap((key) => {
      const value = componentPaths[key];
      return value === undefined ? [] : [[key, value]];
    }),
  );
}

export function buildBaseManifest(
  metadata: PluginMetadata,
  componentPaths: ManifestComponentPaths,
  supportedKeys: readonly ManifestComponentKey[],
): Record<string, unknown> {
  return {
    name: metadata.name,
    version: metadata.version,
    description: metadata.description,
    author: metadata.author,
    license: metadata.license,
    repository: metadata.repository,
    ...(metadata.homepage !== undefined ? { homepage: metadata.homepage } : {}),
    ...(metadata.keywords !== undefined ? { keywords: metadata.keywords } : {}),
    ...pickManifestPaths(componentPaths, supportedKeys),
  };
}

export function camelCaseHooks(
  portable: PortableHooks,
): Record<string, HookEntry[]> {
  const result: Record<string, HookEntry[]> = {};
  for (const event of PORTABLE_EVENTS) {
    const entries = portable[event];
    if (entries !== undefined) result[event] = entries;
  }
  return result;
}

export function pascalMatcherHooks(
  portable: PortableHooks,
): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const event of PORTABLE_EVENTS) {
    const entries = portable[event];
    if (entries === undefined) continue;
    result[PASCAL_MAP[event]] = entries.map((entry) => ({
      matcher: "*",
      hooks: [
        {
          type: entry.type,
          command: entry.command,
          timeout: entry.timeout ?? DEFAULT_HOOK_TIMEOUT,
        },
      ],
    }));
  }
  return result;
}

export function buildBaseAgentArtifact(
  input: AgentArtifactInput,
  agentFrontmatterKeys: ReadonlySet<string>,
): AgentArtifact {
  const { frontmatter, resolvedModel } = input;
  const data: Record<string, unknown> = {
    name: frontmatter.name,
    description: frontmatter.description,
    model: resolvedModel,
  };
  if (frontmatter.tools.length > 0) data.tools = frontmatter.tools;
  for (const key of agentFrontmatterKeys) {
    const value = (frontmatter as unknown as Record<string, unknown>)[key];
    if (value === undefined) continue;
    if (Array.isArray(value) && value.length === 0) continue;
    data[key] = value;
  }
  const appendix =
    agentFrontmatterKeys.has("skills") || frontmatter.skills.length === 0
      ? ""
      : buildSkillsAppendix(frontmatter.skills);
  return { frontmatter: data, appendix };
}

export function buildPortableAgentArtifact(
  input: AgentArtifactInput,
): AgentArtifact {
  return buildBaseAgentArtifact(input, EMPTY_AGENT_KEYS);
}

const EMPTY_AGENT_KEYS: ReadonlySet<string> = new Set();

function buildSkillsAppendix(skills: string[]): string {
  const lines = skills.map((skill) => `- ${skill}`).join("\n");
  return `\n## Required skills (prompt contract)\n\nThis harness does not expose a structured skills binding, so treat the\nfollowing skill names as a hard prompt contract — invoke them by name when\nthe workflow calls for their behavior:\n\n${lines}\n`;
}

File: python/cheese_flow/adapters/_shared.py
"""Port of `src/adapters/_shared.ts` — shared manifest + hook + agent builders."""

from __future__ import annotations

from collections.abc import Iterable
from typing import Any

from cheese_flow.lib.harness import (
    PORTABLE_EVENTS,
    AgentArtifact,
    AgentArtifactInput,
    HookEntry,
    ManifestComponentPaths,
    PluginMetadata,
    PortableHooks,
)

_PASCAL_MAP: dict[str, str] = {
    "sessionStart": "SessionStart",
    "preToolUse": "PreToolUse",
    "postToolUse": "PostToolUse",
}

_DEFAULT_HOOK_TIMEOUT = 600

_EMPTY_AGENT_KEYS: tuple[str, ...] = ()


def _pick_manifest_paths(
    component_paths: ManifestComponentPaths,
    supported_keys: tuple[str, ...],
) -> dict[str, str]:
    return {
        key: component_paths[key]  # type: ignore[literal-required]
        for key in supported_keys
        if key in component_paths and component_paths[key] is not None  # type: ignore[literal-required]
    }


def build_base_manifest(
    metadata: PluginMetadata,
    component_paths: ManifestComponentPaths,
    supported_keys: tuple[str, ...],
) -> dict[str, Any]:
    result: dict[str, Any] = {
        "name": metadata["name"],
        "version": metadata["version"],
        "description": metadata["description"],
        "author": metadata["author"],
        "license": metadata["license"],
        "repository": metadata["repository"],
    }
    homepage = metadata.get("homepage")
    if homepage is not None:
        result["homepage"] = homepage
    keywords = metadata.get("keywords")
    if keywords is not None:
        result["keywords"] = keywords
    result.update(_pick_manifest_paths(component_paths, supported_keys))
    return result


def camel_case_hooks(portable: PortableHooks) -> dict[str, list[HookEntry]]:
    result: dict[str, list[HookEntry]] = {}
    for event in PORTABLE_EVENTS:
        entries = portable.get(event)
        if entries is not None:
            result[event] = entries
    return result


def pascal_matcher_hooks(portable: PortableHooks) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for event in PORTABLE_EVENTS:
        entries = portable.get(event)
        if entries is None:
            continue
        result[_PASCAL_MAP[event]] = [
            {
                "matcher": "*",
                "hooks": [
                    {
                        "type": entry["type"],
                        "command": entry["command"],
                        "timeout": entry.get("timeout", _DEFAULT_HOOK_TIMEOUT),
                    },
                ],
            }
            for entry in entries
        ]
    return result


def _build_skills_appendix(skills: list[str]) -> str:
    lines = "\n".join(f"- {skill}" for skill in skills)
    return (
        "\n## Required skills (prompt contract)\n\n"
        "This harness does not expose a structured skills binding, so treat the\n"
        "following skill names as a hard prompt contract — invoke them by name when\n"
        "the workflow calls for their behavior:\n\n"
        f"{lines}\n"
    )


def build_base_agent_artifact(
    artifact_input: AgentArtifactInput,
    agent_frontmatter_keys: Iterable[str],
) -> AgentArtifact:
    """Build the agent artifact, iterating ``agent_frontmatter_keys`` in order.

    The TS implementation iterates a ``Set`` whose insertion order matches the
    declared key tuple; Python ``frozenset`` is unordered, so callers must pass
    an ordered iterable (tuple/list) to keep the rendered YAML byte-identical.
    """
    keys = tuple(agent_frontmatter_keys)
    key_set: frozenset[str] = frozenset(keys)
    frontmatter = artifact_input.frontmatter
    data: dict[str, Any] = {
        "name": frontmatter["name"],
        "description": frontmatter["description"],
        "model": artifact_input.resolvedModel,
    }
    tools = frontmatter["tools"]
    if len(tools) > 0:
        data["tools"] = tools
    raw: dict[str, Any] = dict(frontmatter)
    for key in keys:
        value = raw.get(key)
        if value is None:
            continue
        if isinstance(value, list) and len(value) == 0:
            continue
        data[key] = value
    skills = frontmatter["skills"]
    appendix = (
        ""
        if "skills" in key_set or len(skills) == 0
        else _build_skills_appendix(skills)
    )
    return AgentArtifact(frontmatter=data, appendix=appendix)


def build_portable_agent_artifact(
    artifact_input: AgentArtifactInput,
) -> AgentArtifact:
    return build_base_agent_artifact(artifact_input, _EMPTY_AGENT_KEYS)
